File: radar_model.py
import numpy as np
import numpy.typing as npt
from typing import Union


class RadarModel:
    # Helper class for simulation of Pd and SNR
    def __init__(
        self,
        snr_50_db: float = 3.768,
        rcs_50: float = 0.854,
        r_50: float = 70,
        pf: float = 1e-2,
        lobe_width_scale_factor: float = 0.2292,
        rcs_min: float = -30,
    ) -> None:
        # The default parameters seems to match the Bosch radar reasonably well

        # Determines the width of the simulated antenna lobe
        self.lobe_width_scale_factor = lobe_width_scale_factor

        # Parameters in linear scale
        self.snr_50 = 10 ** (snr_50_db / 10)
        self.rcs_50 = rcs_50
        self.r_50 = r_50

        self.pf = pf
        assert self.pf < 1

        # RCS drops to this value at 0m
        # Mimics the behavior of the Bosch radar
        self.rcs_min = rcs_min

    def compute_snr(
        self,
        r: Union[npt.NDArray[np.float64], float],
        phi: Union[npt.NDArray[np.float64], float],
        rcs: float,
    ) -> npt.NDArray[np.float64]:
        # Simulate and return the snr

        return (
            self.snr_50
            * (rcs / self.rcs_50)
            * ((self.r_50 / r) ** 4)
            * self.gaussian_gain(phi)
        )

    def gaussian_gain(
        self, phi: Union[npt.NDArray[np.float64], float]
    ) -> npt.NDArray[np.float64]:
        # Antenna lobe, approximated with a gaussian distribution.
        # Matches the figures in the Bosch data sheet reasonably well
        return np.exp(-0.5 * np.square(phi / self.lobe_width_scale_factor))

    def compute_pd(
        self, snr: Union[npt.NDArray[np.float64], float]
    ) -> npt.NDArray[np.float64]:
        # Simulate and return the Pd
        return self.pf ** (1 / (1 + snr))

    def sim_rcs(
        self,
        r: Union[npt.NDArray[np.float64], float],
        rcs: float,
        add_noise: bool = True,
    ) -> npt.NDArray[np.float64]:
        # Simulate the behavior of the Bosch radars RCS estimation
        # RCS is typically reasonably stable (but noisy) at larger distances, but starts
        # to drop sharply as the object gets closer to the radar.
        # It seems to correspond to object size, so the rcs value used for the simulation
        # is used to determine where the RCS should start falling.
        drop_at = rcs * 2
        slope = (rcs - self.rcs_min) / drop_at
        rcs_sim = slope * r + self.rcs_min
        rcs_sim = np.where(r > drop_at, rcs, rcs_sim)
        if add_noise:
            rcs_sim += r * (np.random.random(rcs_sim.shape) + 0.5) / 10
        return rcs_sim

File: src/radarModel.cpp
#include <cmath>
#include "Eigen/Dense"
#include "radarModel.h"

BoschRadar::BoschRadar(const Floating snr50db, const Floating rcs50, const Floating r50, const Floating pf, const Floating scale, const Floating rcs_min) : snr50_(BoschRadar::dbToLinear(snr50db)), rcs50_(rcs50), r50_(r50), pf_(pf), scale_(scale), rcsMin_(rcs_min) {}

Floating BoschRadar::computeSnr(const Floating range, const Floating phi, const Floating rcs)
{
    return snr50_ * (rcs / rcs50_) * std::pow(r50_ / range, 4) * gaussianGain(phi);
}

ArrayXF BoschRadar::computeSnr(const Eigen::Ref<const ArrayXF> &range, const Eigen::Ref<const ArrayXF> &phi, const Floating rcs)
{
    return snr50_ * (rcs / rcs50_) * Eigen::pow(r50_ / range, 4) * gaussianGain(phi);
}

Floating BoschRadar::computePd(const Floating snr)
{
    return std::pow(pf_, 1.0 / (1.0 + snr));
}

ArrayXF BoschRadar::computePd(const Eigen::Ref<const ArrayXF> &snr)
{
    return Eigen::pow(pf_, 1.0 / (1.0 + snr));
}

Floating BoschRadar::gaussianGain(const Floating phi)
{
    const Floating exponent = phi / scale_;
    return std::exp(-0.5f * exponent * exponent);
}

ArrayXF BoschRadar::gaussianGain(const Eigen::Ref<const ArrayXF> &phi)
{
    return (-0.5 * (phi / scale_).square()).exp();
}

Floating BoschRadar::simRcs(const Floating r, const Floating rcs, const bool addNoise)
{
    constexpr Floating randMax = static_cast<Floating>(RAND_MAX);
    const Floating dropAt = rcs * 2.0;
    const Floating noise = addNoise ? r * ((static_cast<Floating>(std::rand()) / randMax) + 0.5) / 10.0 : 0.0;
    if (r > dropAt)
    {
        return rcs + noise;
    }
    else
    {
        const Floating slope = (rcs - rcsMin_) / dropAt;
        return slope * r + rcsMin_ + noise;
    }
}

ArrayXF BoschRadar::simRcs(const Eigen::Ref<const ArrayXF> &r, const Floating rcs, const bool addNoise)
{
    const Floating dropAt = rcs * 2.0;
    const Floating slope = (rcs - rcsMin_) / dropAt;
    ArrayXF rcsSim = slope * r + rcsMin_;
    rcsSim = (r > dropAt).select(rcs, rcsSim);
    if (addNoise)
    {
        rcsSim += r * ((ArrayXF::Random(rcsSim.rows(), rcsSim.cols()) / 2.0) + 1.0) / 10.0;
    }

    return rcsSim;
}
